### PR TITLE
Normalize finding evidence and improve strategy run resilience

### PR DESCRIPTION
**Pull request by aviggiano**
Replays the Ultrafuzz product fixes discovered during a long full-run campaign onto current `main`. These landed on a branch that forked before `main` diverged significantly (evals package, CI TS migration, v0.0.4); this PR carries only the fixes, cleanly rebased with no conflicts, so none of that divergent work is touched.

## Fixes
- **Normalize finding evidence fragments / fields** (`packages/artifacts/src/findings.ts`, template): robustly normalize varied/flexible finding-evidence shapes emitted by agents so downstream rendering and required-artifact checks stay stable.
- **Keep resumed workflow sync state running** (`packages/runtime/src/workflow-sync.ts`): resumed workflows no longer drop their sync state, so a resumed campaign continues instead of stalling.
- **Raise strategy task timeout default** and **give the library differential strategy more time** (`.ultrafuzz/topology.yml`): longer-running strategy nodes no longer time out prematurely.

## Tests
- `pnpm -w build`
- `pnpm --filter @ultrafuzz/artifacts test` (17 pass)
- `pnpm --filter @ultrafuzz/runtime test` (40 pass)
- `pnpm --filter @ultrafuzz/prompts test` (24 pass)
- `pnpm -w format:check`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h3>Greptile Summary</h3>

This PR backports three targeted production fixes: robust normalization of LLM-emitted finding evidence fields (path/fragment/line splitting, command reclassification, `notes` coercion), a sync-state merge correction that prevents resumed workflows from inheriting stale terminal event state, and a group-level `timeout_seconds: 7200` default in the strategy topology. All three are well-covered by new unit and integration tests.

- **Evidence normalization** (`findings.ts`): paths with `#fragment` or `:N` suffixes are now split into canonical `path` + `fragment`/`line` fields; relative paths that look like shell commands are moved from `path` to `command`; bare strings in `notes` are coerced to arrays.
- **Workflow sync** (`workflow-sync.ts`): `mergeNodeWorkflowEvidence` now prefers the current step state (pending/running) over stale terminal events when attempts are equal, fixing resumed campaigns that were stalling; `finalRunStatus` now returns `running` before checking node-level timeout flags, so an actively running workflow isn't mis-classified as timed-out just because a prior node timed out.
- **Topology** (`.ultrafuzz/topology.yml`): raises the per-task default from its previous implicit value to 2 hours, giving long-running strategy nodes headroom.

<h3>Confidence Score: 4/5</h3>

Safe to merge; the changes are well-tested and address real campaign failures. Two minor normalization edge-cases are worth noting but do not affect the happy path.

The core logic changes in both `findings.ts` and `workflow-sync.ts` are covered by new tests that exercise the exact scenarios described in the PR. The two noted issues — greedy regex in `splitLineReference` producing a misleading error for `file:line:col` input, and the `looksLikeEvidenceCommand` heuristic silently reclassifying space-containing paths as commands instead of rejecting them — are both observable only with unusual agent output that rarely (if ever) appears in practice. Neither affects existing passing tests nor the resumed-workflow fix.

`packages/artifacts/src/findings.ts` — specifically the `splitLineReference` regex and `looksLikeEvidenceCommand` path-vs-command boundary.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| packages/artifacts/src/findings.ts | Large expansion of evidence normalization: adds command detection heuristic, path/fragment splitting, line-number parsing, and coerces bare string `notes` to arrays. Logic is generally sound but the greedy line-suffix regex can produce confusing errors for line:column notation. |
| packages/runtime/src/workflow-sync.ts | Two targeted fixes: prefer step state over stale terminal events when attempt numbers are equal (resumed workflows), and check for an actively-running workflow status before falling through to node-level timeout status. Both changes are well-covered by tests. |
| .ultrafuzz/topology.yml | Adds a 2-hour group-level default timeout for all strategy nodes. Simple one-liner change. |
| packages/artifacts/test/artifacts.test.ts | Adds targeted tests for `notes` coercion, command reclassification, path fragment extraction, and line-suffix parsing; conflict-detection paths are also tested. |
| packages/runtime/test/runtime.test.ts | New integration test covers the reset-to-pending case for resumed workflows where events show a prior failure but the step is pending again. |
| .ultrafuzz/prompts/_templates/output-contract/findings.mdx | Adds a single sentence clarifying that `path` fields must be clean relative paths without anchors or line selectors; aligned with the new normalization logic. |
| packages/prompts/test/semantic-anchors.test.ts | Extends the snapshot test to assert the new `without anchors or line selectors` sentence is present in the template. |

</details>

<details><summary><h3>Flowchart</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    subgraph EvidenceNormalization["Evidence Normalization (findings.ts)"]
        A[Evidence entry] --> B{Is string?}
        B -- yes --> C[Pass through as-is]
        B -- no --> D{Is plain object?}
        D -- no --> ERR1[throw]
        D -- yes --> E[Read path field]
        E --> F{path defined and not absolute?}
        F -- no --> G[Skip path normalization]
        F -- yes --> H{looksLikeEvidenceCommand?}
        H -- yes --> I[Move path to command, delete path]
        H -- no --> J[normalizeFindingMetadataPathReference]
        J --> K{Contains hash?}
        K -- yes --> L[Split path + fragment, validate both]
        K -- no --> M{splitLineReference?}
        M -- yes --> N[Split path + line number, validate path]
        M -- no --> O[Validate path only]
        N --> P{line conflicts with existing line?}
        P -- yes --> ERR2[throw]
        P -- no --> Q[Set entry.path, entry.line]
        L --> R{fragment conflicts with existing fragment?}
        R -- yes --> ERR3[throw]
        R -- no --> S[Set entry.path, entry.fragment]
    end
    subgraph WorkflowSync["Sync State Merge (workflow-sync.ts)"]
        W1[fromStep step inspect state] --> W3[mergeNodeWorkflowEvidence]
        W2[fromEvents event log state] --> W3
        W3 --> W4{stepIsTerminal?}
        W4 -- yes --> W8[Events win on status]
        W4 -- no --> W5{eventIsTerminal?}
        W5 -- no --> W8
        W5 -- yes --> W6{event attempt newer than step?}
        W6 -- yes --> W8
        W6 -- no --> W7[Step wins - resume case, step is pending/running]
    end
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    subgraph EvidenceNormalization["Evidence Normalization (findings.ts)"]
        A[Evidence entry] --> B{Is string?}
        B -- yes --> C[Pass through as-is]
        B -- no --> D{Is plain object?}
        D -- no --> ERR1[throw]
        D -- yes --> E[Read path field]
        E --> F{path defined and not absolute?}
        F -- no --> G[Skip path normalization]
        F -- yes --> H{looksLikeEvidenceCommand?}
        H -- yes --> I[Move path to command, delete path]
        H -- no --> J[normalizeFindingMetadataPathReference]
        J --> K{Contains hash?}
        K -- yes --> L[Split path + fragment, validate both]
        K -- no --> M{splitLineReference?}
        M -- yes --> N[Split path + line number, validate path]
        M -- no --> O[Validate path only]
        N --> P{line conflicts with existing line?}
        P -- yes --> ERR2[throw]
        P -- no --> Q[Set entry.path, entry.line]
        L --> R{fragment conflicts with existing fragment?}
        R -- yes --> ERR3[throw]
        R -- no --> S[Set entry.path, entry.fragment]
    end
    subgraph WorkflowSync["Sync State Merge (workflow-sync.ts)"]
        W1[fromStep step inspect state] --> W3[mergeNodeWorkflowEvidence]
        W2[fromEvents event log state] --> W3
        W3 --> W4{stepIsTerminal?}
        W4 -- yes --> W8[Events win on status]
        W4 -- no --> W5{eventIsTerminal?}
        W5 -- no --> W8
        W5 -- yes --> W6{event attempt newer than step?}
        W6 -- yes --> W8
        W6 -- no --> W7[Step wins - resume case, step is pending/running]
    end
```

</a>
</details>

<a href="https://app.greptile.com/ide/claude-code?prompt=Fix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Apackages%2Fartifacts%2Fsrc%2Ffindings.ts%3A329-330%0A**Greedy%20%60%3AN%60%20match%20mis-parses%20%60file%3Aline%3Acol%60%20notation.**%20The%20regex%20in%20%60splitLineReference%60%20is%20greedy%20on%20the%20left%20side%2C%20so%20a%20path%20like%20%60src%2FOracle.sol%3A42%3A3%60%20%28line%2042%2C%20column%203%20%E2%80%94%20emitted%20by%20some%20Foundry%2Fcast%20output%20parsers%29%20is%20parsed%20as%20%60path%20%3D%20%22src%2FOracle.sol%3A42%22%60%20and%20%60line%20%3D%203%60.%20The%20extracted%20path%20then%20fails%20%60validateFindingMetadataRelativePath%60%20with%20an%20%60unsafe-path-segment%60%20error%20because%20%60Oracle.sol%3A42%60%20contains%20%60%3A%60%2C%20producing%20a%20confusing%20message%20that%20gives%20no%20hint%20about%20the%20column%20suffix.%20A%20non-greedy%20path%20group%20makes%20the%20intent%20explicit%20and%20catches%20this%20before%20validation.%0A%0A%60%60%60suggestion%0Afunction%20splitLineReference%28value%3A%20string%29%3A%20%7B%20path%3A%20string%3B%20line%3A%20number%20%7D%20%7C%20undefined%20%7B%0A%20%20const%20match%20%3D%20%2F%5E%28%3F%3Cpath%3E.%2B%3F%29%3A%28%3F%3Cline%3E%5B1-9%5D%5B0-9%5D*%29%24%2Fu.exec%28value%29%3B%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%202%0Apackages%2Fartifacts%2Fsrc%2Ffindings.ts%3A266-274%0A**%60looksLikeEvidenceCommand%60%20silently%20reclassifies%20space-containing%20paths.**%20Any%20relative%20%60path%60%20value%20that%20has%20whitespace%20and%20a%20first%20token%20matching%20%60%5BA-Za-z0-9._%40%2B%2F-%5D%2B%60%20is%20moved%20to%20%60command%60%20rather%20than%20being%20rejected.%20In%20practice%2C%20Solidity%2FFoundry%20paths%20never%20have%20spaces%2C%20but%20if%20an%20agent%20ever%20emits%20%60%7B%20%22path%22%3A%20%22test%20results%2FFoo.t.sol%22%20%7D%60%2C%20the%20entry%20is%20silently%20rewritten%20to%20%60%7B%20%22command%22%3A%20%22test%20results%2FFoo.t.sol%22%20%7D%60%20with%20no%20error.%20Since%20%60validateFindingMetadataRelativePath%60%20would%20reject%20such%20a%20path%20anyway%20%28spaces%20fail%20the%20segment%20regex%29%2C%20the%20only%20observable%20difference%20is%20that%20the%20evidence%20survives%20normalization%20in%20an%20unintended%20shape.%20Consider%20either%20keeping%20the%20path%20as-is%20for%20validation%20rejection%2C%20or%20adding%20an%20explicit%20guard%20that%20%60looksLikeEvidenceCommand%60%20is%20not%20triggered%20when%20the%20value%20resembles%20a%20file%20path%20%28e.g.%2C%20contains%20%60%2F%60%20after%20the%20first%20component%29.%0A%0A&repo=monad-developers%2Fultrafuzz&pr=25&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=6">Fix All in Claude Code</picture></a>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 2 code review issues. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 2
packages/artifacts/src/findings.ts:329-330
**Greedy `:N` match mis-parses `file:line:col` notation.** The regex in `splitLineReference` is greedy on the left side, so a path like `src/Oracle.sol:42:3` (line 42, column 3 — emitted by some Foundry/cast output parsers) is parsed as `path = "src/Oracle.sol:42"` and `line = 3`. The extracted path then fails `validateFindingMetadataRelativePath` with an `unsafe-path-segment` error because `Oracle.sol:42` contains `:`, producing a confusing message that gives no hint about the column suffix. A non-greedy path group makes the intent explicit and catches this before validation.

```suggestion
function splitLineReference(value: string): { path: string; line: number } | undefined {
  const match = /^(?<path>.+?):(?<line>[1-9][0-9]*)$/u.exec(value);
```

### Issue 2 of 2
packages/artifacts/src/findings.ts:266-274
**`looksLikeEvidenceCommand` silently reclassifies space-containing paths.** Any relative `path` value that has whitespace and a first token matching `[A-Za-z0-9._@+/-]+` is moved to `command` rather than being rejected. In practice, Solidity/Foundry paths never have spaces, but if an agent ever emits `{ "path": "test results/Foo.t.sol" }`, the entry is silently rewritten to `{ "command": "test results/Foo.t.sol" }` with no error. Since `validateFindingMetadataRelativePath` would reject such a path anyway (spaces fail the segment regex), the only observable difference is that the evidence survives normalization in an unintended shape. Consider either keeping the path as-is for validation rejection, or adding an explicit guard that `looksLikeEvidenceCommand` is not triggered when the value resembles a file path (e.g., contains `/` after the first component).

`````

</details>

<sub>Reviews (1): Last reviewed commit: ["Normalize flexible finding evidence fiel..."](https://github.com/monad-developers/ultrafuzz/commit/667c26f69c936603f880abafef9386378c83088f) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=43278159)</sub>

> Greptile also left **2 inline comments** on this PR.

### CODE REVIEW

**Comment by aviggiano** [comment]
Smithers verifier approved this PR for merge.

Independently verified PR #25 head branch codex/full-run-product-fixes in its assigned worktree (.worktrees/product-fixes-pr). Working tree clean, HEAD (667c26f) in sync with origin/codex/full-run-product-fixes, origin/main is an ancestor (cleanly rebased), and git merge-tree against origin/main is conflict-free; GitHub reports MERGEABLE (UNSTABLE only from in-progress heavy CI, which is intentionally left to GitHub Actions per validation policy). Diff scope matches the PR body exactly: packages/artifacts/src/findings.ts + findings.mdx template (evidence normalization), packages/runtime/src/workflow-sync.ts (resumed sync state stays running), .ultrafuzz/topology.yml (strategy timeout raises), plus tests in artifacts/runtime/prompts test files — nothing outside scope. Re-checked PR feedback: zero approved-author (aviggiano) comments or reviews; only greptile-apps bot activity, ignored per security policy. Fast targeted checks re-run independently: @ultrafuzz/artifacts 17/17 pass, @ultrafuzz/runtime 40 pass / 0 fail (1 skipped); deterministic validation already passed format:check, prompts tests (24 pass), docs:check, and merge cleanliness. No corrective changes needed. Branch is ready for the workflow to handle push/review/merge once GitHub Actions heavy checks (release-gates, target E2E) complete.

GitHub checks passed for PR #25; merge state is CLEAN.

File: packages/artifacts/src/findings.ts [hunk 1]
```diff
@@ -247,13 +259,115 @@ function validateEvidence(value: unknown, index: number): void {
...
+
+  const relativePath = value.slice(0, hashIndex);
+  const fragment = value.slice(hashIndex + 1);
+  validateFindingMetadataRelativePath(relativePath, key);
+  validateFindingMetadataFragment(fragment, `${key} fragment`);
+  return { path: relativePath, fragment };
+}
+
+function splitLineReference(value: string): { path: string; line: number } | undefined {
+  const match = /^(?<path>.+):(?<line>[1-9][0-9]*)$/u.exec(value);
```
[the remark this answers is not in the document]
Comment (aviggiano) [reply packages/artifacts/src/findings.ts:330]: Accurate observation. A `file:line:col` reference (column suffix) is parsed greedily as path `file:line`, which then fails `validateFindingMetadataRelativePath` because `:` is rejected in path segments. Note this fails safe and is not a regression: before this change any `:`-suffixed path threw the same `unsafe-path-segment` error. The output contract now also instructs agents to keep anchors and line selectors out of `path`. Extending `splitLineReference` to strip an optional `:col` suffix (or improving the error message) is a reasonable follow-up, tracked outside this merged PR.

File: packages/artifacts/src/findings.ts [hunk 2]
```diff
@@ -247,13 +259,115 @@ function validateEvidence(value: unknown, index: number): void {
...
-      validateFindingMetadataRelativePath(evidencePath, "evidence path");
+      if (looksLikeEvidenceCommand(evidencePath)) {
+        const existingCommand = optionalString(entry, "command");
+        if (existingCommand !== undefined && existingCommand !== evidencePath) {
+          throw new Error(`evidence command conflicts with existing command field`);
+        }
+        entry.command = evidencePath;
+        delete entry.path;
+        continue;
+      }
```
[the remark this answers is not in the document]
Comment (aviggiano) [reply packages/artifacts/src/findings.ts:274]: This is the intended tradeoff. A whitespace-containing string whose first token looks like a program name is heuristically indistinguishable from a shell command, and agents in long runs do emit validation commands in the `path` slot. Repository paths in this pipeline never contain spaces (the safe-path segment regex has always rejected them), so before this change such a value hard-failed the entire findings batch; preserving it under `command` keeps the evidence without weakening path validation. If a genuinely space-containing file path ever matters, it would need explicit support in the path validator first.
